From 3f75a3a87958b0e10fbefa9a3483e5ac3f536812 Mon Sep 17 00:00:00 2001
From: Garry Newman <garrynewman@users.noreply.github.com>
Date: Tue, 24 Oct 2023 16:05:30 +0100
Subject: [PATCH] Copy/Paste/Reset transform

---
 .../GameObjectInspector/GameObjectHeader.cs   | 49 +++++++++++++++++++
 1 file changed, 49 insertions(+)

diff --git a/code/Editor/GameObjectInspector/GameObjectHeader.cs b/code/Editor/GameObjectInspector/GameObjectHeader.cs
index 4b37501..544ec63 100644
--- a/code/Editor/GameObjectInspector/GameObjectHeader.cs
+++ b/code/Editor/GameObjectInspector/GameObjectHeader.cs
@@ -2,6 +2,7 @@
 using Editor;
 using Sandbox;
 using System.Linq;
+using System.Numerics;
 
 class GameObjectHeader : Widget
 {
@@ -46,6 +47,54 @@ class GameObjectHeader : Widget
 		Paint.DrawRect( LocalRect );
 	}
 
+	protected override void OnContextMenu( ContextMenuEvent e )
+	{
+		e.Accepted = true;
+
+		var m = new Menu();
+
+		var clipText = EditorUtility.Clipboard.Paste();
+
+		try
+		{
+			var tx = Json.Deserialize<Transform>( clipText );
+			if ( tx != default )
+			{
+				m.AddOption( "Paste Transform", action: () =>
+				{
+					Target.GetProperty( "Transform" ).TryGetAsObject( out var txo );
+					txo.GetProperty( "Local" ).SetValue( tx );
+					SceneEditorSession.Active.FullUndoSnapshot( "Paste Transform" );
+				} );
+
+				m.AddSeparator();
+			}
+		}
+		catch ( System.Exception )
+		{
+			// ignore
+		}
+
+		m.AddOption( "Reset Transform", action: () =>
+		{
+			Target.GetProperty( "Transform" ).TryGetAsObject( out var txo );
+			txo.GetProperty( "Local" ).SetValue( Transform.Zero );
+
+			SceneEditorSession.Active.FullUndoSnapshot( "Reset Transform" );
+		} );
+
+		m.AddSeparator();
+
+		m.AddOption( "Copy Transform", action: () =>
+		{
+			Target.GetProperty( "Transform" ).TryGetAsObject( out var txo );
+			var tx = txo.GetProperty( "Local" ).GetValue<Transform>();
+
+			EditorUtility.Clipboard.Copy( Json.Serialize( tx ) );
+		} );
+
+		m.OpenAtCursor();
+	}
 }
 
 class AttachmentControlWidget : ControlWidget
-- 
2.26.1.windows.1

